man/dodecaphony.man
.TH DODECAPHONY 1 2025-02-08 SHADIZADE

.SH NAME
dodecaphony \- a script to track dodecs in the dodecaphony calendar system

.SH SYNOPSIS
.B dodec -c \fI<command>\fR

.SH DESCRIPTION
This script is meant to track my own 12-day week (dodec) calendar system (dodecaphony), which fits my work schedule better than the historical coincidence that is the 7-day week, an inconvenient prime number. This is not meant to be a general reform, merely a personal system. The dodecaphony is made up of 360 days (30 dodecs), and thus falls short of the Georgian year. Syncronization between the two systems is not important.

.SH COMMANDS
.BR to-dx " " \fI<date>\fR " " \fI[formatting]\fR
.ti 10
Converts a Georgian date to Dodecaphony.
.ti 10
\[bu] Formatting options:
.ti 15
\[bu] \fI4\fR (default): Prints in zz.t.dd.rr format.
.ti 15
\[bu] \fI3\fR:           Prints in zz.dd.rr format.
.ti 15
\[bu] \fI[arbitrary]\fR: replaces %{z,t,d,r} with the corresponding values. Cannot deal with normal spaces (use nbsp).
.ti 12
\fB==> dodec -c 'to-dx 2025-02-10 "Today’s ram is %r."'\fR
.TP
.BR to-dx-char " " \fI<date>\fR " " \fI[:KEYS]\fR
.ti 10
Converts a Georgian date to its corresponding set of characters in Dodecaphony.
.ti 10
\[bu] Keys:
.ti 15
\[bu] \fIcolor\fR: Prints the characters in color.
.ti 12
\fB==> dodec -c 'to-dx-char 2025-02-10 color'\fR
.sp 0
.TP
.BR pretty " " \fI<date>\fR
.ti 10
Converts and prettyprints a Georgian date in Dodecaphony.
.ti 12
\fB==> dodec -c 'pretty 2025-02-10'\fR

.SH SPECIFICATIONS
\[bu] Conversion specifications:
.sp 0
.ti 10
\[bu] 1 \fIram\fR (r)      = 1  Georgian day
.sp 0
.ti 10
\[bu] 1 \fIdodec\fR (d)    = 12 r
.sp 0
.ti 10
\[bu] 1 \fItrithing\fR (t) = 10 d  = 120 r
.sp 0
.ti 10
\[bu] 1 \fIzhou\fR (z)     =  3 t  =  30 d  =  360 r  ≈  1 Georgian year


\- \fIram\fR stands for Ramanujan, who derived the 1+2+3..=–1/12 equation.
.sp 0
\- \fIdodec\fR is short for dodecad, which is any set of twelve.
.sp 0
\- \fItrithing\fR is an old English word meaning one third.
.sp 0
\- \fIzhou\fR 周 in Chinese meaning cycle, and is the surname of notorious workaholic icon Zhou Enlai.
.sp 1
\- The names of the units are never capitalized.
.sp 0
\- Trithings are optional, and they do not reset the dodecs.
.sp 0
.ti 12
\- 01.2.01.01 ⨯ incorrect
.sp 0
.ti 12
\- 01.2.11.01 ✓ correct
.sp 0
.ti 12
\- 01.11.01   ✓ correct
.sp 0
\- The date format is zz.t.dd.rr wherein t is optional.
.sp 0
\- 01.1.01.01 is 2025/02/10.
.sp 1
\- The units of the dodec are each associated with a Chinese character.
.sp 0
.ti 10
\[bu] z1  →  刊 kān     \[bu] t1  →  义 yì     \[bu] d1  →  刃 rèn     \[bu] r1  →  闯 chuǎng  
.sp 0	            
.ti 6	            
                        \[bu] t2  →  而 ér     \[bu] d2  →  技 jì      \[bu] r2  →  岸 àn 
.sp 0	            
.ti 6	            
                        \[bu] t3  →  森 sēn    \[bu] d3  →  凰 huáng   \[bu] r3  →  豹 bào
.sp 0	            
.ti 2	            
                                               \[bu] d4  →  兴 xīng    \[bu] r4  →  斐 fěi
.sp 0	            
.ti 2	            
                                               \[bu] d5  →  琴 qín     \[bu] r5  →  顶 dǐng
.sp 0	            
.ti 2	            
                                               \[bu] d6  →  从 cóng    \[bu] r6  →  卫 wèi 
.sp 0	            
.ti 2	            
                                               \[bu] d7  →  查 chá     \[bu] r7  →  田 tián
.sp 0	            
.ti 2	            
                                               \[bu] d8  →  刷 shuā    \[bu] r8  →  审 shěn
.sp 0	            
.ti 2	            
                                               \[bu] d9  →  汇 huì     \[bu] r9  →  启 qǐ
.sp 0	            
.ti 2	            
                                               \[bu] d10 →  讴 ōu      \[bu] r10 →  剀 kǎi
.sp 0	            
.ti 2	            
                                               \[bu] d11 →  叫 jiào    \[bu] r11 →  矛 máo
.sp 0	            
.ti 2	            
                                               \[bu] d12 →  汝 rǔ      \[bu] r12 →  匝 zā
.sp 0	            
.ti 2	            
                                               \[bu] d13 →  云 yún     
.sp 0	            
.ti 2	            
                                               \[bu] d14 →  罡 gāng
.sp 0	            
.ti 2	            
                                               \[bu] d15 →  功 gōng     
.sp 0	            
.ti 2	            
                                               \[bu] d16 →  更 gèng
.sp 0	            
.ti 2	            
                                               \[bu] d17 →  红 hóng     
.sp 0	            
.ti 2	            
                                               \[bu] d18 →  绕 rào     
.sp 0	            
.ti 2	            
                                               \[bu] d19 →  幺 yāo     
.sp 0	            
.ti 2	            
                                               \[bu] d20 →  套 tào
.sp 0	            
.ti 2	            
                                               \[bu] d21 →  织 zhī     
.sp 0	            
.ti 2	            
                                               \[bu] d22 →  韭 jiǔ     
.sp 0	            
.ti 2	            
                                               \[bu] d23 →  飘 piāo     
.sp 0	            
.ti 2	            
                                               \[bu] d24 →  贝 bèi     
.sp 0	            
.ti 2	            
                                               \[bu] d25 →  庶 shù      
.sp 0	            
.ti 2	            
                                               \[bu] d26 →  忠 zhōng    
.sp 0	            
.ti 2	            
                                               \[bu] d27 →  郭 guō      
.sp 0	            
.ti 2	            
                                               \[bu] d28 →  仍 réng     
.sp 0	            
.ti 2	            
                                               \[bu] d29 →  存 cún     
.sp 0	            
.ti 2	            
                                               \[bu] d30 →  幽 yōu     


dodecaphony/dodecaphony.sh
#!/bin/bash
source ~/Repositories/scripts/essential-functions

function load-global-variables {
    dx_day0="2025-02-09"
    dx_day1="2025-02-10"
    zhou_chars=(z 刊)
    rydyng_chars=(y 义 而 森)
    dodec_chars=(d 刃 技 凰 兴 琴 从 查 刷 汇 讴 叫 汝 云 罡 功 更 红 绕 幺 套 织 韭 飘 贝 庶 忠 郭 仍 存 幽)
    ram_chars=(r 闯 岸 豹 斐 顶 卫 田 审 启 剀 矛 匝)
    zhou_color='38;2;80;118;148'
    rydyng_color='38;2;69;72;206'
    dodec_color='38;2;93;205;217'
    ram_color='38;2;254;248;236'
}   
function load-math-functions {
    function math-ceiling {
	perl -w -e "use POSIX; print ceil($1), qq{\n}"
    }
    function math-floor {
	perl -w -e "use POSIX; print floor($1), qq{\n}"
    }
}


function test-date-validity {
    datetest --isvalid "$@" || return 1
    datetest "$dx_day0" --ot "$@" || return 1
}

function format-dx-string {
    [[ -z "$3" ]] && return 1
    if [[ -z "$4" ]]
    then
	printf "%.2d.%.2d.%.2d" "$1" "$2" "$3"
    else
	printf "%.2d.%.1d.%.2d.%.2d" "$1" "$2" "$3" "$4"
    fi
}

function gx-dx-convert {
    # converts the date from Georgian to Dodecaphony
#    sane "$1"
    test-date-validity "$1" || return 1
    local rams_since_day0="$(datediff "$dx_day0" "$1")"
    # basically epoch for dx
    local ram_in_zhou="$(((rams_since_day0 - 1 ) % 360 + 1))"
    # {1..360}
    local zhous_since_day0="$((rams_since_day0 / 360 + 1))" # bash can't do float, basically floor function
    # z1 begins on dx_day1
    local rydyng_in_zhou="$((ram_in_zhou / 120 + 1))"
    # {1..3}
    local dodec_in_zhou="$(((ram_in_zhou - 1 ) / 12 + 1))"
    # {1..30}
    local ram_in_dodec="$(((ram_in_zhou - 1 ) % 12 + 1 ))"
    # {1..12}
    [[ "$2" = "4" || -z "$2" ]] && {
	format-dx-string "$zhous_since_day0" "$rydyng_in_zhou" "$dodec_in_zhou" "$ram_in_dodec"
	return 0
    }
    [[ "$2" = "3" ]] && {
	format-dx-string "$zhous_since_day0" "$dodec_in_zhou" "$ram_in_dodec"
	return 0
    }
    printf "$(sed "s/z/$zhous_since_day0/g;s/y/$rydyng_in_zhou/g;s/d/$dodec_in_zhou/g;s/r/$ram_in_dodec/g" <<< "$2")"
}

function color-number {
    [[ -z "$1" || -z "$2" ]] && return 1
    case "$1" in
	z) printf "\033[${zhou_color}m%.2d" "${2##0}";;
	y) printf "\033[${rydyng_color}m%.1d" "${2##0}";;
	d) printf "\033[${dodec_color}m%.2d" "${2##0}";;
	r) printf "\033[${ram_color}m%.2d" "${2##0}";;
	*) return 1 ;;
    esac
    printf '\033[0m'
}
    

function test-valid-dx-string {
    [[ -z "$@" ]] && return 1
    grep -q '\.' <<< "$@" || return 1
    grep -q '/' <<< "$@" && return 1
    grep -q '[a-zA-Z ]' <<< "$@" && return 1
    [[ "$(printf "$@" | wc -c)" -gt 10 ]] && return 1
    [[ "$(printf "$@" | wc -c)" -lt 8 ]] && return 1
    [[ "$(printf "$@" | sed 's/[^.]//g' | wc -c)" -gt 3 ]] && return 1
    [[ "$(printf "$@" | sed 's/[^.]//g' | wc -c)" -lt 2 ]] && return 1
    :
}

function no-of-fields-in-dx-string {
    [[ -z "$@" ]] && return 1
    test-valid-dx-string "$@" || return 1
    printf "$@" | awk -F '.' '{print NF}'
}

function extract-from-dx-string {
    [[ -z "$1" || -z "$2" ]] && return 1
    test-valid-dx-string "$2" || return 1
    if [[ "$(no-of-fields-in-dx-string "$2")" -eq 4 ]]
    then
	case "$1" in
	    z) cut -d '.' -f 1 <<< "$2";;
	    y) cut -d '.' -f 2 <<< "$2";;
	    d) cut -d '.' -f 3 <<< "$2";;
	    r) cut -d '.' -f 4 <<< "$2";;
	    *) return 1 ;;
	esac
    else
	case "$1" in
	    z) cut -d '.' -f 1 <<< "$2";;
	    d) cut -d '.' -f 2 <<< "$2";;
	    r) cut -d '.' -f 3 <<< "$2";;
	    *) return 1 ;;
	esac
    fi
}

function truncate-y-from-dx-string {
    local ex_z="$(extract-from-dx-string z "$@")"
    local ex_d="$(extract-from-dx-string d "$@")"
    local ex_r="$(extract-from-dx-string r "$@")"
    printf "$ex_z.$ex_d.$ex_r"
}

function test-valid-dx-string-values {
    :
}

function key-parser {
    [[ -z "$@" ]] && return 1
    keys=(keys ${@/:/ })
}

function give-char {
    [[ -z "$2" || -z "$1" ]] && return 1
    local char_index="$(printf '%.d' "${2##0}")"
    key-parser "$3"
    for j in ${keys[@]}
    do
	case "$j" in
	    "keys") : ;;
	    "color") local zhou_color_sequence="\033[${zhou_color}m"
		     local rydyng_color_sequence="\033[${rydyng_color}m"
		     local dodec_color_sequence="\033[${dodec_color}m"
		     local ram_color_sequence="\033[${ram_color}m";;
	    *) : ;;
	esac
    done
    case "$1" in
	z) printf "$zhou_color_sequence${zhou_chars[$char_index]:-甚}";;
	y) printf "$rydyng_color_sequence${rydyng_chars[$char_index]:-甚}";;
	d) printf "$dodec_color_sequence${dodec_chars[$char_index]:-甚}";;
	r) printf "$ram_color_sequence${ram_chars[$char_index]:-甚}";;
	*) return 1 ;;
    esac
    printf '\033[0m'
}

function give-chars-for-dx-string {
    [[ -z "$1" ]] && return 1
    key-parser "$2"
    for j in ${keys[@]}
    do
	case "$j" in
	    "keys") : ;;
	    "color") local passkey+=":color";;		     
	    *) : ;;
	esac
    done
    local ch_z="$(give-char z $(extract-from-dx-string z "$1" "$passkey"))"
    [[ "$(no-of-fields-in-dx-string "$2")" -eq 3 ]] || {
	local ch_y="$(give-char y $(extract-from-dx-string y "$1" "$passkey"))"
    }
    local ch_d="$(give-char d $(extract-from-dx-string d "$1" "$passkey"))"
    local ch_r="$(give-char r $(extract-from-dx-string r "$1" "$passkey"))"
    printf "$ch_z$ch_y$ch_d$ch_r"
}

function command-option {
    function field {
	cut -d ' ' -f "$2" <<< "$1"
    }
    case "$(field "$@" 1)" in
	"today")
	    gx-dx-convert "$(date +"%Y-%m-%d")" || exit 1
	    echo
	    ;;
	"today-char")
	    give-chars-for-dx-string "$(gx-dx-convert "$(date +"%Y-%m-%d")")" || exit 1
	    echo
	    ;;
	"pretty")
	    test-date-validity "$(field "$@" 2)" || return 1
	    printf "\033[${zhou_color}m∎\033[${rydyng_color}m∎\033[${dodec_color}m∎\033[${ram_color}m∎\033[0m "
	    color-number z "$(gx-dx-convert "$(field "$@" 2)" z)"
	    printf .
	    color-number y "$(gx-dx-convert "$(field "$@" 2)" y)"
	    printf .
	    color-number d "$(gx-dx-convert "$(field "$@" 2)" d)"
	    printf .
	    color-number r "$(gx-dx-convert "$(field "$@" 2)" r)"
	    printf ' '
	    give-chars-for-dx-string "$(gx-dx-convert "$(field "$@" 2)" 4)" color
	    printf '\n'
	    ;;
	"to-dx")
	    gx-dx-convert "$(field "$@" 2)" || exit 1
	    echo
	    ;;
	"to-dx-char")
	    local dx_date="$(gx-dx-convert "$(field "$@" 2)")"
	    give-chars-for-dx-string "$dx_date" "$(field "$@" 3)" || exit 1
	    echo
	    ;;
	"to-dx-no-y")
	    local dx_date="$(gx-dx-convert "$(field "$@" 2)")"
	    truncate-y-from-dx-string "$dx_date" || exit 1
	    echo
	    ;;
	"to-dx-no-y-char")
	    local dx_date="$(gx-dx-convert "$(field "$@" 2)")"
	    local dx_date="$(truncate-y-from-dx-string "$dx_date")"
	    give-chars-for-dx-string "$dx_date" || exit 1
	    echo
	    ;;
	"test")
	    color-number r 3
	    echo
	    ;;
	*)
	    echo ":: Unknown command."
	    exit 1
	    ;;
    esac
}

load-global-variables
load-math-functions
while getopts 'c:' OPTION; do
    case "$OPTION" in
	"c") command-option "$OPTARG" ;;
	*) echo incorrect input; exit ;;
	esac
done
(( $OPTIND == 1 )) && command-option today
# gx-dx-convert "$1"
# echo
# give-chars-for-dx-string $(gx-dx-convert "$1")
# echo

